app/src/main/java/com/example/android/codelabs/paging/data/GithubRemoteMediator.kt

```diff
@@ -0,0 +1,166 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.android.codelabs.paging.data
+
+import android.util.Log
+import androidx.paging.LoadType
+import androidx.paging.PagingState
+import androidx.paging.RemoteMediator
+import androidx.room.withTransaction
+import com.example.android.codelabs.paging.api.GithubService
+import com.example.android.codelabs.paging.api.IN_QUALIFIER
+import com.example.android.codelabs.paging.db.RemoteKeys
+import com.example.android.codelabs.paging.db.RepoDatabase
+import com.example.android.codelabs.paging.model.Repo
+import retrofit2.HttpException
+import java.io.IOException
+import java.io.InvalidObjectException
+
+// GitHub page API is 1 based: https://developer.github.com/v3/#pagination
+const val GITHUB_STARTING_PAGE_INDEX = 1
+
+class GithubRemoteMediator(
+        private val query: String,
+        private val service: GithubService,
+        private val repoDatabase: RepoDatabase
+) : RemoteMediator<Int, Repo>() {
+
+    /**
+     * The paging library will call this method when there is no more data in the database to
+     * load from.
+     *
+     * For every item we retrieve we store the previous and next keys, so we know how to request
+     * more data, based on that specific item.
+     *
+     * state.pages holds the pages that were loaded until now and saved in the database.
+     *
+     * If the load type is REFRESH it means that we are in one of the following situations:
+     *  * it's the first time when the query was triggered and there was nothing saved yet in the
+     *  database
+     *  * a swipe to refresh was called
+     * If the load type is START, we need to rely on the first item available in the pages
+     * retrieved so far to compute they key for the previous page.
+     * If the load type is END, we need to rely on the last item available in the pages
+     * retrieved so far to compute they key for the next page.
+     */
+    override suspend fun load(loadType: LoadType, state: PagingState<Int, Repo>): MediatorResult {
+        val page = when (loadType) {
+            LoadType.REFRESH -> {
+                val remoteKeys = getRemoteKeyClosestToCurrentPosition(state)
+                remoteKeys?.nextKey?.minus(1) ?: GITHUB_STARTING_PAGE_INDEX
+            }
+            LoadType.START -> {
+                val remoteKeys = getRemoteKeyForFirstItem(state)
+                if (remoteKeys == null) {
```

Comment:
You could replace this with a `checkNotNull` which _might_ be more idiomatic, it throws IllegalStateException which might be more apt here though.

Reply:
Although this will be obsolete once yigit's change lands.

https://android-review.googlesource.com/c/platform/frameworks/support/+/1295842

Might want to just wait for this.